bash_completionsV2.go

```diff
@@ -0,0 +1,241 @@
+    __%[1]s_debug "The completions are: ${out[*]}"
+
+    if [ $((directive & shellCompDirectiveError)) -ne 0 ]; then
+        # Error code.  No completion.
+        __%[1]s_debug "Received error from custom completion go code"
+        return
+    else
+        if [ $((directive & shellCompDirectiveNoSpace)) -ne 0 ]; then
+            if [[ $(type -t compopt) = "builtin" ]]; then
+                __%[1]s_debug "Activating no space"
+                compopt -o nospace
+            fi
+        fi
+        if [ $((directive & shellCompDirectiveNoFileComp)) -ne 0 ]; then
+            if [[ $(type -t compopt) = "builtin" ]]; then
+                __%[1]s_debug "Activating no file completion"
+                compopt +o default
+            fi
+        fi
+    fi
+
+    if [ $((directive & shellCompDirectiveFilterFileExt)) -ne 0 ]; then
+        # File extension filtering
+        local fullFilter filter filteringCmd
+
+        # Do not use quotes around the $out variable or else newline
+        # characters will be kept.
+        for filter in ${out[*]}; do
+            fullFilter+="$filter|"
+        done
+
+        filteringCmd="_filedir $fullFilter"
+        __%[1]s_debug "File filtering command: $filteringCmd"
+        $filteringCmd
+    elif [ $((directive & shellCompDirectiveFilterDirs)) -ne 0 ]; then
+        # File completion for directories only
+
+        # Use printf to strip any trailing newline
+        local subdir=$(printf "%%s" "${out[0]}")
+        if [ -n "$subdir" ]; then
+            __%[1]s_debug "Listing directories in $subdir"
+            pushd "$subdir" >/dev/null 2>&1 && _filedir -d && popd >/dev/null 2>&1 || return
+        else
+            __%[1]s_debug "Listing directories in ."
+            _filedir -d
+        fi
+    else
+        local tab=$(printf '\t')
+        local longest=0
+        # Look for the longest completion so that we can format things nicely
+        while IFS='' read -r comp; do
+            comp=${comp%%%%$tab*}
+            if ((${#comp}>$longest)); then
+                longest=${#comp}
+            fi
+        done < <(printf "%%s\n" "${out[@]}")
+
+        local completions=()
+        while IFS='' read -r comp; do
+            if [ -z "$comp" ]; then
+                continue
+            fi
+
+            __%[1]s_debug "Original comp: $comp"
+
+            # First properly format the description string which follows a tab character if there is one
+            if [[ "$comp" == *$tab* ]]; then
+                desc=${comp#*$tab}
+                comp=${comp%%%%$tab*}
+
+                # Add the proper number of spaces to align the descriptions
+                for ((i = ${#comp} ; i < $longest ; i++)); do
+                    comp+=" "
+                done
+                comp+="  ($desc)"
+
+                # Now escape all spaces
+                comp=${comp// /\\ }
+
+                # Also escape single quotes
+                comp=${comp//\'/\\\'}
```

Comment:
You also need to escape the backticks as they can lead to arbitrary code execution and will break the script.

```
comp=${comp//\`/\\\`}
```

Also it would make sense to make sure it doesn't double escape if the command description already contains a escaped version like "\\`" or "\\'"

Reply:
Good point for the backticks, I have now escaped them.  I've also checked that they don't need escaping in zsh or fish.

As for the double escaping, can you give an example of Go code where this could happen?  It is unclear to me how the escaping backslash would be retained after Go has printed the string.

Reply:
Basically if the cmd/flag description contains something like this:
`rootCmd.Flags().String("escape", "", "description with space\\ test\\ end")`

`./testprog __complete -` will print this:
`--escape        description with space\ test\ end`
but the bash completion shows only this:
`--escape  (description with space\`

This happen because we convert "`\ `" to "`\\ `" in the bash script, which than fails because the space is now unescaped.

I know it very unlikely that a description contains something like that but it wouldn't hurt if we took care of it.

Reply:
Oh I just realized we just would need to replace the backslash with backslashes before we escape the other special characters.

Reply:
I just found out that the bash-builtin `printf` can handle the escaping for us.
From `man bash-builtins`:
`%q     causes printf to output the corresponding argument in a format that can be reused as shell input.`

For Example:
```
$ printf "%q" "hello'"
hello\'
$ printf "%q" 'hello`'
hello\`
$ printf "%q" " "
\      <- there is a space here
```

So we just would need to run `printf -v comp "%q"  "${comp}"` and all special characters will be escaped.
I believe this would be the best option to handle the escaping problem.

Reply:
@Luap99 Nice trick with the `printf`!  I actually hit a problem with double quotes and was going to need to escape those as well, but with `printf %q`, everything is escaped for me.  Thanks!